lib/charms/grafana_agent_k8s/v0/grafana_agent.py

```diff
@@ -155,37 +192,88 @@ def _on_log_proxy_relation_departed(self, event):
         self._update_agents_list(event)
 
         if len(self._current_config["clients"]) == 0:
-            self._container.stop(self._container_name)
+            self._container.stop("promtail")
         else:
-            self._container.restart(self._container_name)
+            self._container.restart("promtail")
 
     def _on_upgrade_charm(self, event):
         # TODO: Implement it ;-)
         pass
 
-    def _on_promtail_pebble_ready(self, _):
-        """Event handler for the pebble ready event.
+    def _get_container(self, container_name):
+        if container_name is not None:
+            return self._charm.unit.get_container(container_name)
 
-        Args:
-            event: The event object of the pebble ready event
-        """
+        containers = dict(self._charm.model.unit.containers)
+
+        if len(containers) == 1:
+            return self._charm.unit.get_container([*containers].pop())
+
+        # FIXME: Use custom exception
+        raise Exception("Container cannot be obtained")
 
+    def _build_pebble_layer(self):
+        """Event handler for the pebble ready event."""
         pebble_layer = {
             "summary": "promtail layer",
             "description": "pebble config layer for promtail",
             "services": {
                 "promtail": {
                     "override": "replace",
                     "summary": "promtail",
-                    "command": f"/usr/bin/promtail {self._cli_args}",
-                    "startup": "disable",
+                    "command": "{} {}".format(WORKLOAD_BINARY_PATH, self._cli_args),
+                    "startup": "enabled",
                 }
             },
         }
+        self._container.add_layer(self._container_name, pebble_layer, combine=True)
+
+    def _obtain_promtail(self, event) -> None:
+        if self._is_promtail_binary_in_workload():
+            return
+
+        self._download_promtail(event)
+
+        if not self._check_sha256sum():
```

Comment:
I'd maybe just combine `_download_promtail` and `self._check_sha256sum` and just have the `_download_promtail` method raise the `PromtailDigestError` if the checksum is off? It seems the `_check_sha256sum` is not called anywhere else so probably not that useful in its own right?

Reply:
Yes, we are not using `_chech_sha256sum` outside this scope. Makes sense. Thanks.